docs: allow TOC navigation

ulysses4ever opened the pull request:
It annoys me beyond measure that I can't freely browse TOC (in the left sidebar) without jumping between pages. It turned out there's a little switch that allows you to (**almost**) freely unfold menu items, and I turn it on in this PR. 

Why "almost" is because it will still only show one section unfolded at a time. There's a long-standing PR to fix that: https://github.com/readthedocs/sphinx_rtd_theme/pull/692 I think that it's still way better than the current state.

Another related but different idea is to have the TOC fully unfolded always. I'm less sure that I'd want that (perhaps I'd), but `sphinx_rtd_theme` currently doesn't allow it out of the box: https://github.com/readthedocs/sphinx_rtd_theme/issues/455

Here's what TOC will look like, and you can unfold entries by clicking in [+] without leaving the current page
![2022-06-07_20-06-1654647339](https://user-images.githubusercontent.com/6832600/172505052-0430c71a-6ecf-4ad1-87c4-b8cea9914597.png)

ulysses4ever commented:
Hmm, it doesn't seem to work [when rendered on RTD]()

andreasabel commented:
> Hmm, it doesn't seem to work [when rendered](https://cabal--8196.org.readthedocs.build/en/8196/)… It did work locally.

Have you checked in the [build report](https://readthedocs.org/api/v2/build/17109381.txt) that the same package versions are used locally and on rtd?

In case of mismatch, you can upgrade requirements in 
https://github.com/haskell/cabal/blob/5b64eae1e797690b988b434bdce39f1c919aed7d/doc/requirements.in#L1-L5
and then run `make -C doc` to generate `requirements.txt`.
E.g. recently we upgraded to `Sphinx >= 5` and `sphinx_rtd_theme >= 1` for the Agda docs:
- https://github.com/agda/agda/pull/5943

andreasabel commented:
I added a commit that upgrades to Sphinx >= 5.

This seems to trigger [new warnings](https://github.com/haskell/cabal/runs/6776955142?check_suite_focus=true#step:4:102):
> WARNING: extlinks: Sphinx-6.0 will require a caption string to contain exactly one '%s' and all other '%' need to be escaped as '%%'.

In the rendered docs, I still do not see a change of behavior (over 3.6) what the TOC concerns.

ulysses4ever commented:
Mystery solved: https://github.com/readthedocs/readthedocs.org/issues/1622 i.e. RTD server messes around with `conf.py`, but there's a workaround, which I added to this patch. I also added a screenshot in the original post for busy reviewers! Otherwise, the preview is [at the same place](https://cabal--8196.org.readthedocs.build/en/8196/) -- now with the changes in effect!

I'm not sure what to do with the warnings introduced with Sphinx update: as we learned, Sphinx version is nor related to the mystery, and the warnings make CI red...

andreasabel commented:
> I'm not sure what to do with the warnings introduced with Sphinx update:

In my last commit, I am trying to fix this by following the instruction on https://www.sphinx-doc.org/en/master/usage/extensions/extlinks.html

ulysses4ever commented on doc/conf.py:
```diff
@@ -81,6 +81,13 @@
+if on_rtd:  # only import and set the theme if we're building docs locally
+    html_theme = 'sphinx_rtd_theme'
+    html_style = None
+    using_rtd_theme = True
```
Great catch, thank you! Just did it.

ulysses4ever commented:
I intend to reshuffle and squash manually such that there are two commits: one about Sphinx update, and one about TOC unfolding. Let me know if you have a better idea.

fishtreesugar commented:
👍🏻  https://cabal.readthedocs.io/en/latest looks amazing!
